packages/components/src/theme/ThemeProvider.test.tsx

```diff
@@ -0,0 +1,149 @@
+    setThemePreloadData: jest.fn(),
+  };
+});
+
+const customThemes = [{ themeKey: 'themeA' }] as [ThemeData];
+const preloadA: ThemePreloadData = { themeKey: 'themeA' };
+
+beforeEach(() => {
+  jest.clearAllMocks();
+  jest.restoreAllMocks();
+  expect.hasAssertions();
+
+  asMock(calculatePreloadStyleContent)
+    .mockName('calculatePreloadStyleContent')
+    .mockReturnValue(':root{mock-preload-content}');
+
+  asMock(getThemePreloadData).mockName('getThemePreloadData');
+  asMock(setThemePreloadData).mockName('setThemePreloadData');
+});
+
+describe('ThemeProvider', () => {
+  const themeContextValueRef = { current: null as ThemeContextValue | null };
+
+  function MockChild() {
+    themeContextValueRef.current = useTheme();
+    return <div>Child</div>;
+  }
+
+  beforeEach(() => {
+    themeContextValueRef.current = null;
+  });
+
+  it.each([
+    [null, null],
+    [null, preloadA],
+    [customThemes, null],
+    [customThemes, preloadA],
+  ] as const)(
+    'should load themes based on preload data or default: %s, %s',
+    (themes, preloadData) => {
+      asMock(getThemePreloadData).mockReturnValue(preloadData);
+
+      const component = render(
+        <ThemeProvider themes={themes}>
+          <MockChild />
+        </ThemeProvider>
+      );
+
+      assertNotNull(themeContextValueRef.current);
+
+      expect(themeContextValueRef.current.activeThemes).toEqual(
+        themes == null
+          ? null
+          : getActiveThemes(preloadData?.themeKey ?? DEFAULT_DARK_THEME_KEY, {
+              base: getDefaultBaseThemes(),
+              custom: themes,
+            })
+      );
+
+      expect(themeContextValueRef.current.selectedThemeKey).toEqual(
+        preloadData?.themeKey ?? DEFAULT_DARK_THEME_KEY
+      );
+
+      expect(component.baseElement).toMatchSnapshot();
+    }
+  );
+
+  it.each([
+    [null, null],
+    [null, preloadA],
+    [customThemes, null],
+    [customThemes, preloadA],
+  ] as const)(
+    'should set preload data when active themes change: %s, %s',
+    (themes, preloadData) => {
+      asMock(getThemePreloadData).mockReturnValue(preloadData);
+
+      render(
+        <ThemeProvider themes={themes}>
+          <MockChild />
+        </ThemeProvider>
+      );
+
+      expect(setThemePreloadData).toHaveBeenCalledWith({
+        themeKey: preloadData?.themeKey ?? DEFAULT_DARK_THEME_KEY,
+        preloadStyleContent: calculatePreloadStyleContent(),
+      });
+    }
+  );
+
+  describe.each([null, customThemes])('setSelectedThemeKey: %s', themes => {
+    it.each([DEFAULT_LIGHT_THEME_KEY, customThemes[0].themeKey])(
+      'should change selected theme: %s',
+      themeKey => {
+        const component = render(
+          <ThemeProvider themes={themes}>
+            <MockChild />
+          </ThemeProvider>
+        );
+
+        assertNotNull(themeContextValueRef.current);
+
+        act(() => {
+          themeContextValueRef.current!.setSelectedThemeKey(themeKey);
```

Comment:
```suggestion
          themeContextValueRef.current?.setSelectedThemeKey(themeKey);
```

Reply:
In this particular case I know it's not null due to the `assertNotNull` above. It's just the fact that `act` runs a callback, that the typesystem is unaware. Seems like good case for null assertion to me

Reply:
Since it's a test that's fair enough. Either way if it is actually null the test should and will fail.

TS is right about the possibility of null here though. The ref value _could_ be changed from the time the assertion is done to the time the callback is run (TS doesn't know the callback should be immediately invoked by `act`)

Reply:
Agreed, TS is right about the general possibility of a ref changing before a callback is executed, but this would be a case where the dev knows more about the runtime particulars, namely that act runs its callback synchronously when called (at least that's my understanding of how `act` works). I believe this is the type of use case `!` is intended for, "dev knows better, so ignore this case". That said, if my understanding of how `act` works is incorrect, I can change it.

Reply:
I'm pretty sure act is synchronous, so it's fine. I believe we have the linter rule turned off in tests b/c there are a few cases where we needed it or it made sense to test some failure cases or something.

I've always seen `!` as less needed now that optional chaining and nullish coaelescing exist. The non-null assertion was added in TS 2.0 which was 2016. Optional chaining and nullish coaelescing made their way into browsers in early 2020.

Fine to leave it as is though